File: 1014 Trie Tree/main.cpp
#include <iostream>
#include <map>
#include <string>

using namespace std;

class TrieTreeNode
{
  public:
    map<char, TrieTreeNode *> children;
    int count;

    TrieTreeNode()
    {
        this->count = 0;
    }
};

class TrieTree
{
  public:
    TrieTreeNode *root;

    TrieTree()
    {
        this->root = new TrieTreeNode;
    }

    void insert(string word)
    {
        auto current= this->root;
        for (auto c : word)
        {
            auto finding = current->children.find(c);
            if (finding == current->children.end())
            {
                current->children[c] = new TrieTreeNode();
            }
            current = current->children[c];
            current->count += 1;
        }
    }

    int query(string word)
    {
        auto current = this->root;
        for (auto c : word)
        {
            auto finding = current->children.find(c);
            if (finding == current->children.end())
            {
                return 0;
            }
            current = current->children[c];
        }
        return current->count;
    }

};

int main()
{
    int m, n;
    string s;

    auto trie = TrieTree();

    cin >> m;
    for (int i = 0; i < m; ++i)
    {
        cin >> s;
        trie.insert(s);
    }

    cin >> n;
    for (int i = 0; i < n; ++i)
    {
        cin >> s;
        cout << trie.query(s) << endl;
    }
}

File: 1014 Trie Tree/main.py
from __future__ import print_function, division, absolute_import, unicode_literals


class TrieTreeNode:
    def __init__(self, count=0):
        self.children = {}
        self.count = count


class TrieTree:
    def __init__(self):
        self.root = TrieTreeNode()

    def insert(self, word):
        current = self.root

        for c in word:
            if c not in current.children:
                current.children[c] = TrieTreeNode()
            current = current.children[c]
            current.count += 1

    def query(self, word):
        current = self.root
        for c in word:
            if c not in current.children:
                return 0
            current = current.children[c]
        return current.count


if __name__ == '__main__':
    trie = TrieTree()

    m = int(raw_input())
    for _ in range(m):
        trie.insert(raw_input())

    n = int(raw_input())
    for _ in range(n):
        print(trie.query(raw_input()))
